### PR TITLE
Migrations: Use reflection order for columns in CreateTable

### LINKED ISSUE #2272: Column ordering
**Outcome from much discussion on this topic...**

In EF6 we tried having column order match the order of properties in the class. The issue is that reflection can return a different order on different architectures. Although the order of properties in the model does not theoretically matter... it actually created a lot of very difficult to debug issues on EF6 where the order had a subtle effect on conventions, model diffing, etc. We managed to fix some of them (after many releases) but there are still some issues we have never really fixed. For that reason, metadata is stored in a deterministic (alphabetical) order in metadata in EF7.

In EF6, we also had a column ordering API, but this only impacts the initial create of the table. If you later change the order, migrations does not handle this (it would require a table rebuild). Also, if you add new columns, they are always appended to the table (since anything else also requires a table rebuild).

We could, however, store an ordering hint as an annotation on the model (based on the reflection order). This would _only_ affect the order that migrations writes the columns in a create table call. That way it doesn't matter if it changes when you run your app on different platforms, since it's just an annotation that is ignored for the most part. If we wanted to, we could also flow an order that was specified in `[Column]` annotations into the hint field.

### PR DESCRIPTION
**Pull request by bricelam**
Fixes #2272

### CODE REVIEW

File: src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs [hunk 1]
```diff
@@ -605,6 +606,117 @@ protected virtual IEnumerable<MigrationOperation> Remove([NotNull] string source
...
+                types.GetOrAddNew(clrType)[index] = clrProperty;
+            }
+
+            foreach (var group in unorderedGroups)
+            {
+                groups.Add(group.Key, group.Value.Values.ToList());
+            }
+
+            foreach (var definingForeignKey in entityType.GetDeclaredReferencingForeignKeys()
+                .Where(fk => fk.DeclaringEntityType.DefiningEntityType == entityType))
```
Comment (bricelam) [inline src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs]: @AndriySvyryd is this this right? In Relational, do navigation-defined entity types always map to the principal entity type's table?

Comment (AndriySvyryd) [reply src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs]: No, that's just the default.

Comment (bricelam) [reply src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs]: K, will add a test and react

**Comment by bricelam** [comment]
The more code I write, the less confident I become. 😄 I'm doing one last round of bashing on this before I merge...

**Pull request merged**